jibby.go

```diff
@@ -0,0 +1,529 @@
+// it will be stripped.  Because only UTF-8 is supported, other BOMs are errors.
+// This function consumes leading white space and checks if the first character
+// is '['.  If so, the input format is expected to be a single JSON array of
+// objects and the stream will consist of the objects in the array.  Any read
+// error (including io.EOF) during these checks will be returned.
+//
+// If the the bufio.Reader's size is less than 8192, it will be rebuffered.
+// This is necessary to account for lookahead for long decimals to minimize
+// copying.
+func NewDecoder(json *bufio.Reader) (*Decoder, error) {
+	if json.Size() < 8192 {
+		json = bufio.NewReaderSize(json, 8192)
+	}
+	err := handleBOM(json)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &Decoder{
+		json:     json,
+		maxDepth: 200,
+	}
+
+	ch, err := d.readAfterWS()
+	if err != nil {
+		// Before an object is read, EOF is a valid response that
+		// shouldn't be wrapped.
+		if err == io.EOF {
+			return nil, err
+		}
+		return nil, newReadError(err)
+	}
+
+	switch ch {
+	case '[':
+		d.arrayStarted = true
+	default:
+		_ = d.json.UnreadByte()
+	}
+
+	return d, err
+}
+
+// ExtJSON toggles whether extended JSON is interpreted by the decoder.
+// See https://docs.mongodb.com/manual/reference/mongodb-extended-json/index.html
+// Jibby has limited support for the legacy extended JSON format.
+func (d *Decoder) ExtJSON(b bool) {
+	d.extJSONAllowed = b
+}
+
+// MaxDepth sets the maximum allowed depth of a JSON object.  The default is
+// 200.
+func (d *Decoder) MaxDepth(n int) {
+	d.maxDepth = n
+}
+
+// Decode converts a single JSON object from the input stream into BSON object.
+// The function takes an output buffer as an argument.  If the buffer is not
+// large enough, a new buffer will be allocated when needed.  The final buffer
+// is returned, just like with `append`.  The function returns io.EOF if no
+// objects remain in the stream.
+func (d *Decoder) Decode(buf []byte) ([]byte, error) {
+	if d.arrayFinished {
+		return nil, io.EOF
+	}
+
+	ch, err := d.readAfterWS()
+	if err != nil {
+		// Before an object is read, EOF is a valid response that
+		// shouldn't be wrapped.
+		if err == io.EOF {
+			return nil, err
+		}
+		return nil, newReadError(err)
+	}
+
+	switch ch {
+	case '{':
+		_ = d.json.UnreadByte()
+	case ']':
+		// This case will only occur for an empty top-level array: `[]`.
+		// Otherwise, the closing array bracket is read after an object.
+		if d.arrayStarted {
+			d.arrayFinished = true
+			return nil, io.EOF
+		}
+		return nil, d.parseError(ch, "Decode only supports object decoding")
+	default:
+		return nil, d.parseError(ch, "Decode only supports object decoding")
+	}
+
+	buf, err = d.convertValue(buf, topContainer)
+	if err != nil {
+		return nil, err
+	}
+
+	// If in comma mode, consume comma or ']', otherwise, put the
+	// the character back to be
+	// After ']', terminate stream?
+	if d.arrayStarted {
```

Comment:
I'm not following how we can get to this point. It looks like you set `arrayStarted` in `NewDecoder` if you read a `'['`, but then the first thing you do here is read the next character. If it's anything other than `'{'` or `']'`, you return an error. Is this dead code?